File: frontend/js/views/solar_payback.js
// Solar payback — net cost after credits, payback years, lifetime savings, and
// ROI, via /calc/solar-payback. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }) + '%';
const yrs = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }));

export async function renderSolarPayback(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.solar.h1.title">// SOLAR PAYBACK</span></h1>
        <p class="muted small" data-i18n="view.solar.hint.intro">
            When a rooftop solar system pays for itself. Net cost is the install price less the 30%
            federal credit and any incentives; annual savings rise with utility-rate inflation but
            fall as panels degrade. Shows the payback year, lifetime savings, and ROI. Updates as
            you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.solar.h2.inputs">The system</h2>
            <form id="solar-form" class="inline-form">
                <label><span data-i18n="view.solar.label.cost">System cost ($)</span>
                    <input type="number" step="0.01" min="0" name="system_cost_usd" value="30000" required></label>
                <label><span data-i18n="view.solar.label.fed">Federal credit (%)</span>
                    <input type="number" step="0.1" min="0" name="federal_credit_pct" value="30" required></label>
                <label><span data-i18n="view.solar.label.incentives">Other incentives ($)</span>
                    <input type="number" step="0.01" min="0" name="other_incentives_usd" value="0"></label>
                <label><span data-i18n="view.solar.label.savings">Year-1 savings ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_savings_usd" value="2000" required></label>
                <label><span data-i18n="view.solar.label.deg">Annual degradation (%)</span>
                    <input type="number" step="0.1" min="0" name="annual_degradation_pct" value="0.5"></label>
                <label><span data-i18n="view.solar.label.inf">Electricity inflation (%)</span>
                    <input type="number" step="0.1" min="0" name="electricity_inflation_pct" value="3"></label>
                <label><span data-i18n="view.solar.label.horizon">Horizon (years)</span>
                    <input type="number" step="1" min="1" name="horizon_years" value="25"></label>
            </form>
        </div>
        <div id="solar-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#solar-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            system_cost_usd: Number(fd.get('system_cost_usd')) || 0,
            federal_credit_pct: Number(fd.get('federal_credit_pct')) || 0,
            other_incentives_usd: Number(fd.get('other_incentives_usd')) || 0,
            annual_savings_usd: Number(fd.get('annual_savings_usd')) || 0,
            annual_degradation_pct: Number(fd.get('annual_degradation_pct')) || 0,
            electricity_inflation_pct: Number(fd.get('electricity_inflation_pct')) || 0,
            horizon_years: Number(fd.get('horizon_years')) || 0,
        };
        try {
            const r = await api.calcSolarPayback(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.solar.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#solar-result');
    const profitCls = r.net_profit_usd >= 0 ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.solar.h2.result">The return</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.solar.card.payback">Payback (years)</div>
                    <div class="value pos">${yrs(r.payback_years)}</div></div>
                <div class="card"><div class="label" data-i18n="view.solar.card.net">Net cost</div>
                    <div class="value">${money(r.net_cost_usd)}</div></div>
                <div class="card ${profitCls}"><div class="label" data-i18n="view.solar.card.roi">Lifetime ROI</div>
                    <div class="value ${profitCls}">${pct(r.roi_pct)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.solar.row.credit">Federal credit</td><td>${money(r.federal_credit_usd)}</td></tr>
                    <tr><td data-i18n="view.solar.row.net">Net cost</td><td>${money(r.net_cost_usd)}</td></tr>
                    <tr><td data-i18n="view.solar.row.lifetime">Lifetime savings</td><td>${money(r.lifetime_savings_usd)}</td></tr>
                    <tr><td data-i18n="view.solar.row.payback">Payback (years)</td><td>${yrs(r.payback_years)}</td></tr>
                    <tr class="emph ${profitCls}"><td data-i18n="view.solar.row.profit">Net profit</td><td>${money(r.net_profit_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/solar_payback.rs
//! Solar payback — when a rooftop PV system pays for itself and the lifetime
//! return.
//!
//! Net cost is the install price less the 30% federal credit and any other
//! incentives. Annual savings (avoided electricity) rise with utility inflation
//! but fall as panels degrade:
//!
//! ```text
//! net cost     = system cost × (1 − federal credit) − incentives
//! year y saving = savings₁ × (1 + inflation)^(y−1) × (1 − degradation)^(y−1)
//! payback       = year cumulative savings first cover net cost
//! ```

use serde::{Deserialize, Serialize};

fn d_fed() -> f64 {
    30.0
}
fn d_deg() -> f64 {
    0.5
}
fn d_inf() -> f64 {
    3.0
}
fn d_horizon() -> f64 {
    25.0
}

#[derive(Debug, Clone, Deserialize)]
pub struct SolarInput {
    pub system_cost_usd: f64,
    /// Federal tax credit, percent of system cost.
    #[serde(default = "d_fed")]
    pub federal_credit_pct: f64,
    /// Other rebates/incentives, dollars.
    #[serde(default)]
    pub other_incentives_usd: f64,
    /// First-year electricity savings.
    pub annual_savings_usd: f64,
    /// Annual panel output degradation, percent.
    #[serde(default = "d_deg")]
    pub annual_degradation_pct: f64,
    /// Annual utility-rate inflation, percent.
    #[serde(default = "d_inf")]
    pub electricity_inflation_pct: f64,
    /// Analysis horizon in years (system life).
    #[serde(default = "d_horizon")]
    pub horizon_years: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct SolarResult {
    /// System cost after the federal credit and incentives.
    pub net_cost_usd: f64,
    pub federal_credit_usd: f64,
    /// Total savings over the horizon.
    pub lifetime_savings_usd: f64,
    /// Years until cumulative savings cover the net cost; `None` if never.
    pub payback_years: Option<f64>,
    /// Lifetime savings − net cost.
    pub net_profit_usd: f64,
    /// Net profit / net cost, percent.
    pub roi_pct: f64,
}

pub fn analyze(input: &SolarInput) -> SolarResult {
    let credit = input.system_cost_usd * input.federal_credit_pct / 100.0;
    let net_cost = (input.system_cost_usd - credit - input.other_incentives_usd).max(0.0);
    let inf = input.electricity_inflation_pct / 100.0;
    let deg = input.annual_degradation_pct / 100.0;
    let years = input.horizon_years.max(0.0) as i64;

    let mut cum = 0.0;
    let mut lifetime = 0.0;
    let mut payback = None;
    for y in 1..=years {
        let s = input.annual_savings_usd
            * (1.0 + inf).powi((y - 1) as i32)
            * (1.0 - deg).powi((y - 1) as i32);
        let prev = cum;
        cum += s;
        lifetime += s;
        if payback.is_none() && prev < net_cost && cum >= net_cost && s != 0.0 {
            payback = Some((y - 1) as f64 + (net_cost - prev) / s);
        }
    }
    // Net cost could be zero (incentives ≥ cost) → instant payback.
    if payback.is_none() && net_cost <= 0.0 {
        payback = Some(0.0);
    }

    SolarResult {
        net_cost_usd: net_cost,
        federal_credit_usd: credit,
        lifetime_savings_usd: lifetime,
        payback_years: payback,
        net_profit_usd: lifetime - net_cost,
        roi_pct: if net_cost > 0.0 {
            (lifetime - net_cost) / net_cost * 100.0
        } else {
            0.0
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> SolarInput {
        SolarInput {
            system_cost_usd: 30_000.0,
            federal_credit_pct: 30.0,
            other_incentives_usd: 0.0,
            annual_savings_usd: 2_000.0,
            annual_degradation_pct: 0.5,
            electricity_inflation_pct: 3.0,
            horizon_years: 25.0,
        }
    }

    #[test]
    fn net_cost_after_credit() {
        let r = analyze(&base());
        assert!(close(r.federal_credit_usd, 9_000.0));
        assert!(close(r.net_cost_usd, 21_000.0));
    }

    #[test]
    fn lifetime_savings() {
        assert!(close(analyze(&base()).lifetime_savings_usd, 68_182.96));
    }

    #[test]
    fn payback() {
        assert!(close(analyze(&base()).payback_years.unwrap(), 9.4422));
    }

    #[test]
    fn profit_and_roi() {
        let r = analyze(&base());
        assert!(close(r.net_profit_usd, 47_182.96));
        assert!(close(r.roi_pct, 224.6808));
    }

    #[test]
    fn flat_case_simple_payback() {
        // No inflation/degradation: 21,000 / 2,000 = 10.5 years.
        let r = analyze(&SolarInput {
            annual_degradation_pct: 0.0,
            electricity_inflation_pct: 0.0,
            ..base()
        });
        assert!(close(r.payback_years.unwrap(), 10.5));
        assert!(close(r.lifetime_savings_usd, 50_000.0));
    }

    #[test]
    fn incentives_lower_net_cost() {
        let r = analyze(&SolarInput {
            other_incentives_usd: 5_000.0,
            ..base()
        });
        assert!(close(r.net_cost_usd, 16_000.0));
        assert!(r.payback_years.unwrap() < analyze(&base()).payback_years.unwrap());
    }

    #[test]
    fn never_pays_back_short_horizon() {
        // 3-year horizon can't recover a 21k net cost at ~2k/yr.
        let r = analyze(&SolarInput {
            horizon_years: 3.0,
            ..base()
        });
        assert!(r.payback_years.is_none());
        assert!(r.net_profit_usd < 0.0);
    }

    #[test]
    fn higher_inflation_speeds_payback() {
        let low = analyze(&base());
        let high = analyze(&SolarInput {
            electricity_inflation_pct: 8.0,
            ..base()
        });
        assert!(high.payback_years.unwrap() < low.payback_years.unwrap());
    }
}
